notification/notification.go

```diff
@@ -1,8 +1,72 @@
 package notification
 
-import "github.com/flanksource/incident-commander/api"
+import (
+	"encoding/json"
+	"time"
 
-type Notifier interface {
-	// NotifyResponderAdded sends notifications about a new responder on an incident.
-	NotifyResponderAdded(ctx *api.Context, responder api.Responder) error
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/incident-commander/api"
+	"github.com/patrickmn/go-cache"
+)
+
+var (
+	notificationByIDCache = cache.New(time.Hour*1, time.Hour*1)
+
+	// a separate cache because we purge the caches in two different ways.
+	notificationByEventCache = cache.New(time.Hour*1, time.Hour*1)
+)
+
+func PurgeCache(notificationID string) {
+	notificationByEventCache.Flush()
+	notificationByIDCache.Delete(notificationID)
+}
+
+func GetNotifications(ctx *api.Context, eventName string) ([]models.Notification, error) {
+	if val, found := notificationByEventCache.Get(eventName); found {
+		return val.([]models.Notification), nil
+	}
+
+	var notifications []models.Notification
+	if err := ctx.DB().Where("deleted_at IS NULL").Where("? = ANY(events)", eventName).Find(&notifications).Error; err != nil {
+		return nil, err
+	}
+	notificationByEventCache.Set(eventName, notifications, cache.DefaultExpiration)
```

Comment:
We should compile the filter expressions and put them in the cache, so that they are not recompiled for every event.

Reply:
all expressions are cached by default in `utils/expression.go`